Implement filesystem check

Blajda opened the pull request:
# Description
Implementation of the filesystem check operation.

The implementation is fairly straight forward with a HEAD call being made for each active file to check if it exists.
A remove action is then made for each file that is orphaned.

An alternative solution is instead to maintain a hashset with all active files and then recursively list all files. If the file exists then remove from the set. All remaining files in the set are then considered orphaned.
 
Looking for feedback and if the second approach is preferred I can make the changes

# Related Issue(s)
- closes #1092

houqp commented:
Instead of heading each individual file, wouldn't it be more efficient to issue an prefixed list call to gather all current files in the table? This way, we can reduce API call count by about 1000x.

Blajda commented:
@houqp I modified the process to instead use a list call but I'm not quite sure on how I would take advantage of prefixes in this case. For partitioned tables it could perform a list per partition path but I think non-partitioned tables would still require a full list on the root.

For non-partitioned tables would it be possible to perform a list without including the files in `_delta_log`?

houqp commented:
@Blajda the prefix filter can only be applied with the table root unfortunately, so it will always include the _detal_log folder unfortunately. The main benefit it provides is we can restrict the listing to the table we are running the operation on, not all other tables stored in the same bucket.

houqp commented:
@roeap @wjones127 want to take a final look?

wjones127 commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,174 @@
+//! Audit the Delta Table for active files that do not exist in the underlying filesystem and remove them
+//!
```
A few additional details might be helpful.

```suggestion
//! Audit the Delta Table for active files that do not exist in the underlying filesystem and remove them.
//!
//! Active files are ones that have an add action in the log, but no corresponding remove action.
//! This operation creates a new transaction containing a remove action for each of the missing files.
//!
//! This can be used to repair tables where a data file has been deleted accidentally or 
//! purposefully, if the file was corrupted.
```

wjones127 commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,174 @@
+            });
+        }
+
+        let mut actions = Vec::new();
+        let mut removed_file_paths = Vec::new();
```
```suggestion
        let mut actions = Vec::with_capacity(self.files_to_remove.len());
        let mut removed_file_paths = Vec::with_capacity(self.files_to_remove.len());
```

wjones127 commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,174 @@
+            }));
+        }
+
+        if !actions.is_empty() {
```
This should never be false, right? Since we already return early above if `self.files_to_remove` is empty, and `actions` should have the same length as that, right?

wjones127 commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,174 @@
+                    FileSystemCheckMetrics {
+                        files_removed: plan
+                            .files_to_remove
+                            .iter()
+                            .map(|f| f.path.clone())
```
I think you can consume the plan with `into_iter()` and just take ownership of the plan:
```suggestion
                            .into_iter()
                            .map(|f| f.path)
```

wjones127 commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,174 @@
+
+impl FileSystemCheckBuilder {
+    /// Create a new [`FileSystemCheckBuilder`]
+    pub fn new(store: Arc<DeltaObjectStore>, state: DeltaTableState) -> Self {
```
Should we validate that the provided state is the latest version of the table? Since I don't think this is a valid operation for earlier states? or maybe that is an overall issue with the operations module?

Blajda replied:
I think dry run of this operation would be helpful for diagnosing issues that arise from a combination of vacuum + time travel. If executed on a older version I do expect it to fail during the commit operation.
I can add a test to demonstrate that.

wjones127 commented on rust/tests/command_filesystem_check.rs:
```diff
@@ -0,0 +1,112 @@
+#![cfg(all(feature = "integration_test"))]
```
Thanks for adding integration tests!

wjones127 commented on rust/tests/command_filesystem_check.rs:
```diff
@@ -0,0 +1,112 @@
+#[serial]
+async fn test_filesystem_check_aws() -> TestResult {
+    set_env_if_not_set("AWS_S3_ALLOW_UNSAFE_RENAME", "true");
+    set_env_if_not_set("AWS_S3_LOCKING_PROVIDER", "none");
```
Could be wrong, but I *think* you only need the first environment variable. If that's not the case, we might want to file an issue for that.
```suggestion
```

Blajda replied:
Yes it's required. I had a commit  with just allow unsafe set and it failed since http is not supported for the lock client. Currently the lock client is always constructed and is not determined by `UNSAFE_RENAME`

See: https://github.com/delta-io/delta-rs/blob/main/rust/src/storage/s3.rs#L402

wjones127 commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,174 @@
+            files_to_check.insert(active.path.to_owned(), active);
+        });
+
+        let mut files = self.store.list(None).await?;
```
Delta Tables are allowed to have data files outside the table root. Currently, if that is the case, this operation will remove them all from the log. I see two options:

1. Call HEAD on each file outside the root to check if they exist.
2. Error if we detect any files outside the root.

Blajda replied:
Makes sense. Do any examples of tables that use absolute paths? 
I just checked the protocol and noticed absolute paths are encoded as defined in [rfc2396](https://www.ietf.org/rfc/rfc2396.txt)

It states the following
```
An absolute URI contains the name of the scheme being used (<scheme>)
   followed by a colon (":") and then a string (the <scheme-specific-
   part>) whose interpretation depends on the scheme.
```

So I can segment paths based of it the scheme is set on the path and then use the correct api call for each.

wjones127 replied:
Oh that's a good point! Checking for the scheme at the beginning makes sense.

We don't have any example tables right now. We don't generally support them in most operations, but we should try to keep them in mind. Hard to have example tables without setting up a public S3 bucket or something.

Blajda replied:
I'm not comfortable with supporting absolute paths without having some integration tests. Currently the operation will fail when an absolute path is encountered.

wjones127 reviewed (changes requested):
Thank you for adding this! I've been thinking having an operation like this would let me sleep better at night :)

I think we should address the case where data files are outside the root, either explicitly not supporting or using HEAD for those ones. Also noticed a few minor issues.

houqp commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,174 @@
+
+    async fn create_fsck_plan(&self) -> DeltaResult<FileSystemCheckPlan> {
+        let mut files_to_remove = Vec::new();
+        let mut files_to_check: HashMap<String, &Add> = HashMap::new();
+        let version = self.state.version();
+        let store = self.store.clone();
+
+        self.state.files().iter().for_each(|active| {
+            files_to_check.insert(active.path.to_owned(), active);
+        });
```
yeah, have you tried `HashMap<&str, &Add>`?

Blajda replied:
Yes I works. Updated to avoid the clones on the path.

wjones127 commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    async fn create_fsck_plan(&self) -> DeltaResult<FileSystemCheckPlan> {
+        let mut files_to_remove = Vec::new();
```
```suggestion
```

Blajda commented:
@wjones127 I updated the PR to fail when an absolute path exists in the Delta log. Also added an additional test to demonstrate the operation will only work on the latest table version.

roeap commented on rust/src/operations/filesystem_check.rs:
```diff
@@ -0,0 +1,221 @@
+    pub files_to_remove: Vec<Add>,
+}
+
+fn is_absolute_path(path: &str) -> DeltaResult<bool> {
+    match Url::parse(path) {
+        Ok(_) => Ok(true),
+        Err(ParseError::RelativeUrlWithoutBase) => Ok(false),
+        Err(_) => Err(DeltaTableError::Generic(format!(
+            "Unable to parse path: {}",
+            &path
+        ))),
+    }
+}
```
I think this will not work for local paths. One pattern we apply in `object_store` and here is to try and cannonicalize the path and if it fails proceed with url parsing. e.g.

https://github.com/delta-io/delta-rs/blob/a60129bb48a24f94e8312c12940f6ec96685af05/rust/src/builder.rs#L355-L358

Blajda replied:
I don't follow on why it would not work on local paths. Paths in the Delta log must follow [rfc2396](https://www.ietf.org/rfc/rfc2396.txt) which states absolute paths must have a scheme.
Do you have a particular example that would cause the unit test for this function to fail?

roeap replied:
well  .. in that case I am wrong! If that paths are required to have a scheme, then url parsing should always work for valid paths ... The windows case is also covered in your tests, so all is well :).

roeap reviewed (commented):
Thanks @Blajda  - one small question along the lines of hoping to eventually support absolute paths in the delta log. Otherwise looks great!
